dodiscover/toporder/das.py

```diff
@@ -0,0 +1,127 @@
+import numpy as np
+from numpy.typing import NDArray
+from scipy.stats import ttest_ind
+
+from dodiscover.toporder._base import SteinMixin
+from dodiscover.toporder.score import SCORE
+from dodiscover.toporder.utils import fullAdj2Order
+
+
+class DAS(SCORE):
+    """The Discovery At Scale (DAS) algorithm :footcite:`Montagna2023a` for causal discovery.
+
+    The method infer the topological ordering using SCORE :footcite:`rolland2022`.
+    Then it prunes the fully connected DAG by inspection of the non diagonal entries of the
+    Hessian of the log likelihood.
+    A final, computationally cheap, pruning step is performed with
+    CAM pruning :footcite:`Buhlmann2013`.
+    The method assumes Additive Noise Model and Gaussianity of the noise terms.
+    DAS is a highly scalable method, allowing to run causal discovery on thousands of nodes.
+    It reduces the computational complexity of the pruning method of an order of magnitude with
+    respect to SCORE.
+
+    Parameters
+    ----------
+    eta_G: float
+        Regularization parameter for Stein gradient estimator
+    eta_H : float
+        Regularization parameter for Stein Hessian estimator
+    cam_cutoff : float
+        alpha value for independence testing for CAM edge pruning
+    das_cutoff : float
+        alpha value for hypothesis testing in preliminary DAS pruning
+    n_splines : int
+        Number of splines to use for feature selection with GAM
+        (Generalized Additive Model) fitting.
+        Automatically decreased in case of insufficient samples
+    splines_degree: int
+        Order of splines for feature selection with GAM (Generalized Additive Model) fitting.
+    min_parents : int
+        Minimum number of edges retained by DAS preliminary pruning step.
+        min_parents < 5 doesn't significantly affects execution time, while increasing the accuracy.
+    max_parents : int
+        Maximum number of parents allowed for a single node.
+        Given that CAM pruning is inefficient for > ~20 nodes, larger values are not advised.
+        The value of max_parents should be decrease under the assumption of sparse graphs.
+    """
+
+    def __init__(
+        self,
+        eta_G: float = 0.001,
+        eta_H: float = 0.001,
+        cam_cutoff: float = 0.001,
+        das_cutoff: float = 0.01,
+        n_splines: int = 10,
+        splines_degree: int = 3,
+        min_parents: int = 5,
+        max_parents: int = 20,
+    ):
+        super().__init__(
+            eta_G, eta_H, cam_cutoff, n_splines, splines_degree, estimate_variance=True, pns=False
+        )
+        self.min_parents = min_parents
+        self.max_parents = max_parents
+        self.das_cutoff = das_cutoff
+
+    def prune(self, X: NDArray, A_dense: NDArray) -> NDArray:
```

Comment:
You don't access `self` here.  Make it a static method?

Reply:
Thanks for pointing that. Actually I was creating a `_SteinMixin` instance in the method, even though it is an ancestor of `DAS`. So now I fixed that by accessing `_SteinMixin` methods using `self`